# PISA 2012 DATA CLEANING

In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv('pisa2012.csv/pisa2012.csv', encoding='latin-1')
df.head()

C:\Users\jsamala\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3018: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD       NC  SCHOOLID  \
0           1  Albania     80000  ALB0006  Non-OECD  Albania         1   
1           2  Albania     80000  ALB0006  Non-OECD  Albania         1   
2           3  Albania     80000  ALB0006  Non-OECD  Albania         1   
3           4  Albania     80000  ALB0006  Non-OECD  Albania         1   
4           5  Albania     80000  ALB0006  Non-OECD  Albania         1   

   STIDSTD  ST01Q01  ST02Q01   ...     W_FSTR75  W_FSTR76 W_FSTR77 W_FSTR78  \
0        1       10      1.0   ...      13.7954   13.9235  13.1249  13.1249   
1        2       10      1.0   ...      13.7954   13.9235  13.1249  13.1249   
2        3        9      1.0   ...      12.7307   12.7307  12.7307  12.7307   
3        4        9      1.0   ...      12.7307   12.7307  12.7307  12.7307   
4        5        9      1.0   ...      12.7307   12.7307  12.7307  12.7307   

   W_FSTR79 W_FSTR80 WVARSTRR VAR_UNIT SENWGT_STU  VER_STU  
0    4.3389  13.0829       19        1     0.2098  22NOV13  
1    4.3389  13.0829       19        1     0.2098  22NOV13  
2    4.2436  12.7307       19        1     0.1999  22NOV13  
3    4.2436  12.7307       19        1     0.1999  22NOV13  
4    4.2436  12.7307       19        1     0.1999  22NOV13  

[5 rows x 636 columns]

In [3]:
df.shape

(485490, 636)

This dataset contains information from 485,490 students and 646 features.

Due to having such large amount of features, I wish to limit the number of features.


For my analysis, I wanted to focus on how gender plays a role in not only students' mathematics scores but also students' attitude towards mathematics. I also wanted to take a look at if the parents attitude towards children's mathematics differ among the gender. For these reasons, I have decided to focus on the following features:

**Student Information**
+ CNT: Country
+ ST04Q01: Gender


**Student Mathematics Scores**
+ PV1MATH:	Overall Mathematics Score


**Student's Attitude Towards Mathematics**
+ **Instrumental Motivation**
    + ST29Q02:	Worthwhile for Work
    + ST29Q05:	Worthwhile for Career Chances
    + ST29Q07:	Important for Future Study
    + ST29Q08:	Helps to Get a Job


+ **Math Anxiety**
    + ST42Q01:	Worry That It Will Be Difficult
    + ST42Q03:	Get Very Tense
    + ST42Q05:	Get Very Nervous
    + ST42Q08:	Feel Helpless
    + ST42Q10:	Worry About Getting Poor <Grades>
	
+ **Math Self-Concept**
    + ST42Q02:	Not Good at Maths
    + ST42Q04:	Get Good (Grades)
    + ST42Q06:	Learn Quickly
    + ST42Q07:	One of Best Subjects
    + ST42Q09:	Understand Difficult Work

+ **Math Interest**
    + ST29Q01:	Enjoy Reading
    + ST29Q03:	Look Forward to Lessons
    + ST29Q04:	Enjoy Maths
    + ST29Q06:	Interested
    
+ **Math Work Ethic**	
    + ST46Q01:	Homework Completed in Time
    + ST46Q02:	Work Hard on Homework
    + ST46Q03:	Prepared for Exams
    + ST46Q04:	Study Hard for Quizzes
    + ST46Q05:	Study Until I Understand Everything
    + ST46Q06:	Pay Attention in Classes
    + ST46Q07:	Listen in Classes
    + ST46Q08:	Avoid Distractions When Studying
    + ST46Q09:	Keep Work Organized
    
+ **Math Behaviour**	
    + ST49Q01:	Talk about Maths with Friends
    + ST49Q02:	Help Friends with Maths
    + ST49Q03:	(Extracurricular) Activity
    + ST49Q04:	Participate in Competitions
    + ST49Q05:	Study More Than 2 Extra Hours a Day
    + ST49Q06:	Play Chess
    + ST49Q07:	Computer programming
    + ST49Q09:	Participate in Math Club


**Subjective Norms**
   + ST35Q04:	Parents Believe Studying Mathematics Is Important
   + ST35Q05:	Parents Believe Mathematics Is Important for Career
   + ST35Q06:	Parents Like Mathematics


## Data Assessment

**To-Do Notes:**
+ create more descriptive feature labels


In [4]:
df.duplicated().any()

False

In [5]:
#organize features by category list

#CNT: country, ST04Q01: gender
student_info=['CNT','ST04Q01']

#PV1MATH: Overall Math Score
scores=['PV1MATH']

#ST29Q02: Worthwhile for Work, ST29Q05: Worthwhile for Career Chances, ST29Q07: Important for Future Study
#ST29Q08: Helps to Get a Job
motivation=['ST29Q02','ST29Q05','ST29Q07','ST29Q08']

#ST42Q01: Worry That It Will Be Difficult, ST42Q03: Get Very Tense, ST42Q05: Get Very Nervous
#ST42Q08: Feel Helpless, ST42Q10: Worry About Getting Poor <Grades>
anxiety=['ST42Q01','ST42Q03','ST42Q05','ST42Q08','ST42Q10']

#ST42Q02: Not Good at Maths, ST42Q04: Get Good (Grades), ST42Q06: Learn Quickly
#ST42Q07: One of Best Subjects, ST42Q09: Understand Difficult Work
self=['ST42Q02','ST42Q04','ST42Q06','ST42Q07','ST42Q09']

#ST29Q01: Enjoy Reading, ST29Q03: Look Forward to Lessons, ST29Q04: Enjoy Maths, ST29Q06: Interested
interest=['ST29Q01','ST29Q03','ST29Q04','ST29Q06']

#ST46Q01: Homework Completed in Time, ST46Q02: Work Hard on Homework, ST46Q03: Prepared for Exams
#ST46Q04: Study Hard for Quizzes, ST46Q05: Study Until I Understand Everything, ST46Q06: Pay Attention in Classes
#ST46Q07: Listen in Classes, ST46Q08: Avoid Distractions When Studying, ST46Q09: Keep Work Organized
work_ethic= ['ST46Q01','ST46Q02','ST46Q03','ST46Q04','ST46Q05','ST46Q06','ST46Q07','ST46Q08','ST46Q09']

#ST49Q01: Talk about Maths with Friends,ST49Q02: Help Friends with Maths, ST49Q03: (Extracurricular) Activity
#ST49Q04: Participate in Competitions, ST49Q05: Study More Than 2 Extra Hours a Day, ST49Q06: Play Chess
#ST49Q07: Computer programming, ST49Q09: Participate in Math Club
behavior=['ST49Q01','ST49Q02','ST49Q03','ST49Q04','ST49Q05','ST49Q06','ST49Q07','ST49Q09']

#ST35Q04: Parents Believe Studying Mathematics Is Important
#ST35Q05: Parents Believe Mathematics Is Important for Career
#ST35Q06:Parents Like Mathematics
parents=['ST35Q04','ST35Q05','ST35Q06']

## STUDENT INFO: Country, Gender

In [6]:
df['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

**Looks like some are incorrectly labeled as a country when is a state or city (ie. Floria (USA))**

http://www.oecd.org/pisa/aboutpisa/pisa-2012-participants.htm

**The above link informs us that they have a total of 65 participating countries.**

**Since our data is indicting 68 unique values under countries, this means that 3 values are not correct. Good thing we can easily see (at the bottom of our value count list) which 3 values are incorrectly labels! They are 'Florida (USA)', 'Massachusetts (USA)', 'Connecticut (USA)'.**

In [7]:
df['CNT'].value_counts()[['United States of America','Florida (USA)','Massachusetts (USA)','Connecticut (USA)']].sum()

10294

**There should be a total of 10,294 students under 'United States of America'**

In [8]:
df[student_info].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 2 columns):
CNT        485490 non-null object
ST04Q01    485490 non-null object
dtypes: object(2)
memory usage: 7.4+ MB


**No missing values for countries and gender**

In [9]:
df['ST04Q01'].value_counts()

Female    245064
Male      240426
Name: ST04Q01, dtype: int64

**Looks like gender is almost evenly split**

## Mathematics Scores

In [10]:
df[scores].head()

PV1MATH
0  406.8469
1  486.1427
2  533.2684
3  412.2215
4  381.9209

In [11]:
df[scores].describe()

PV1MATH
count  485490.000000
mean      469.621653
std       103.265391
min        19.792800
25%       395.318600
50%       466.201900
75%       541.057800
max       962.229300

In [12]:
df[scores].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 1 columns):
PV1MATH    485490 non-null float64
dtypes: float64(1)
memory usage: 3.7 MB


**We have no missing values under students' overall score (PV1MATH)**

## Student's Attitude Towards Mathematics

In [13]:
#list of features
attitudes=[motivation,anxiety,self,interest,work_ethic,behavior]

In [14]:
#finds the different categories under each feature
i=0
for _ in attitudes:
    i+=1
    for feature in _:
        print(i,'\n',df[feature].value_counts(),'\n')

1 
 Agree                150479
Strongly agree        97133
Disagree              49792
Strongly disagree     18069
Name: ST29Q02, dtype: int64 

1 
 Agree                153313
Strongly agree        99425
Disagree              44218
Strongly disagree     18275
Name: ST29Q05, dtype: int64 

1 
 Agree                132131
Strongly agree        90470
Disagree              65853
Strongly disagree     26612
Name: ST29Q07, dtype: int64 

1 
 Agree                150347
Strongly agree        85236
Disagree              57949
Strongly disagree     21700
Name: ST29Q08, dtype: int64 

2 
 Agree                134489
Disagree              83359
Strongly agree        68696
Strongly disagree     27311
Name: ST42Q01, dtype: int64 

2 
 Disagree             135714
Agree                 80655
Strongly disagree     60345
Strongly agree        35462
Name: ST42Q03, dtype: int64 

2 
 Disagree             140694
Agree                 84811
Strongly disagree     54941
Strongly agree        32178
Name: ST

**Features under Motivation, Anxiety, Self, Interest, Work_Ethic contain categories Strongly Agree, Agree, Disagree and Strongly Agree.**


**Features under Behavior contains categories Always or Almost Always, Often, Sometimes, Never or Rarely**

In [15]:
df[motivation].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 4 columns):
ST29Q02    315473 non-null object
ST29Q05    315231 non-null object
ST29Q07    315066 non-null object
ST29Q08    315232 non-null object
dtypes: object(4)
memory usage: 14.8+ MB


**We have some missing values here**

In [16]:
df[motivation].isnull().any(axis=1).sum()

172796

**About 172,796 students have at least one feature under motivation left Null**

In [17]:
df[motivation].isnull().sum(axis=1).value_counts()

0    312694
4    168774
1      2822
3       640
2       560
dtype: int64

**A total of 168,774 students left all of these answers Null under Motivation**


**The question is what should we do about these students when calculating Motivation as well as our other features (that describe students' attitudes) that have a Null value?**


In [18]:
df[anxiety].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 5 columns):
ST42Q01    313855 non-null object
ST42Q03    312176 non-null object
ST42Q05    312624 non-null object
ST42Q08    312456 non-null object
ST42Q10    312853 non-null object
dtypes: object(5)
memory usage: 18.5+ MB


In [19]:
df[self].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 5 columns):
ST42Q02    313502 non-null object
ST42Q04    311980 non-null object
ST42Q06    312327 non-null object
ST42Q07    312583 non-null object
ST42Q09    312223 non-null object
dtypes: object(5)
memory usage: 18.5+ MB


**Note to self: The self section asks students whether they believe they are:**

**1) Not good at Math**

**2) Get Good (Grades)**

**3) Learn Quickly**

**4) One of Best Subjects**

**5) Understand Difficult Work**

**Becareful when assigning point values due to fact that "Not Good at Math" is sometime negative while the remainder are positive. In this case, for 'Not good at Math' (ST42Q02) we would assign 'Strongly Agree' a point value of 1, 'Agree' a point value of 2 and so on.**

In [20]:
df[interest].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 4 columns):
ST29Q01    315911 non-null object
ST29Q03    314928 non-null object
ST29Q04    314737 non-null object
ST29Q06    314746 non-null object
dtypes: object(4)
memory usage: 14.8+ MB


In [21]:
df[work_ethic].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 9 columns):
ST46Q01    313898 non-null object
ST46Q02    313567 non-null object
ST46Q03    312994 non-null object
ST46Q04    312997 non-null object
ST46Q05    313043 non-null object
ST46Q06    312900 non-null object
ST46Q07    312854 non-null object
ST46Q08    312989 non-null object
ST46Q09    313040 non-null object
dtypes: object(9)
memory usage: 33.3+ MB


In [22]:
df[behavior].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 8 columns):
ST49Q01    313495 non-null object
ST49Q02    313025 non-null object
ST49Q03    312168 non-null object
ST49Q04    312378 non-null object
ST49Q05    312582 non-null object
ST49Q06    312571 non-null object
ST49Q07    312425 non-null object
ST49Q09    312752 non-null object
dtypes: object(8)
memory usage: 29.6+ MB


In [23]:
#Calculates the number of null values for each section 
#as well as percent of students who answered all questions under each section
i=0
for _ in attitudes:
    i+=1
    print(i)
    d=dict(df[_].isnull().sum(axis=1).value_counts())
    print(d)
    print('% Complete: {}'.format(d[0]/485490))
    print('\n')

1
{0: 312694, 4: 168774, 1: 2822, 3: 640, 2: 560}
% Complete: 0.6440791777379554


2
{0: 307425, 5: 170577, 1: 5846, 2: 733, 4: 562, 3: 347}
% Complete: 0.6332262250509795


3
{0: 306931, 5: 170736, 1: 5949, 2: 939, 4: 523, 3: 412}
% Complete: 0.6322086963686173


4
{0: 311611, 4: 168590, 1: 3999, 3: 699, 2: 591}
% Complete: 0.6418484417804692


5
{0: 306731, 9: 170312, 1: 5393, 2: 1207, 8: 677, 3: 452, 4: 235, 5: 223, 6: 134, 7: 126}
% Complete: 0.6317967414364869


6
{0: 307112, 8: 170977, 1: 5158, 2: 843, 7: 666, 3: 228, 4: 187, 6: 175, 5: 144}
% Complete: 0.6325815155821953




**For each section, at least 60% of students answer all the questions**


**To-Do List:**
  + If student did not answer at least one feature under each section, then fill in Null Value for section's overall score

## SUBJECTIVE NORMS: Parents

In [24]:
#ST35Q04: Parents Believe Studying Mathematics Is Important
#ST35Q05: Parents Believe Mathematics Is Important for Career
#ST35Q06:Parents Like Mathematics

In [25]:
df[parents].head()

ST35Q04         ST35Q05 ST35Q06
0           Agree           Agree   Agree
1           Agree        Disagree   Agree
2  Strongly agree  Strongly agree   Agree
3             NaN             NaN     NaN
4  Strongly agree  Strongly agree   Agree

In [26]:
for _ in parents:
    print(_)
    print(df[_].value_counts(),'\n')

ST35Q04
Agree                153693
Strongly agree       132309
Disagree              23316
Strongly disagree      5842
Name: ST35Q04, dtype: int64 

ST35Q05
Agree                144854
Strongly agree       115112
Disagree              46278
Strongly disagree      8599
Name: ST35Q05, dtype: int64 

ST35Q06
Agree                148876
Disagree              93913
Strongly agree        49488
Strongly disagree     21112
Name: ST35Q06, dtype: int64 



**This section is categorized by:**
   + Strongly Agree
   + Agree
   + Disagree
   + Strongly Agree
   
**We would want to combine these features by assiging point values and taking its mean.**


In [27]:
df[parents].isnull().any(axis=1).sum(), df[parents].isnull().any(axis=1).sum()/df.shape[0]

(173808, 0.35800531421862447)

**About 35.8% of students had left at least one of these questions blank**
   + When combining scores under parent's section, we should leave these 35.8% null

### DATA ASSESSMENT SUMMARY

+ Delete unneccessary features


+ Rename feature labels


+ Correctly categorize countries under CNT (country): Replace values of 'Florida (USA)','Massachusetts (USA)', and 'Connecticut (USA)' with 'United States of America'


+ Student and Parent's Attitudes: Combine the features in each section into one feature by assigning point values and taking average.
  + If student did not answer at least one feature under each section, then fill in Null Value for section's overall score
  + Becareful when assigning point values due to fact that "Not Good at Math" is sometime negative while the remainder are positive. In this case, for 'Not good at Math' (ST42Q02) we would assign 'Strongly Agree' a point value of 1, 'Agree' a point value of 2 and so on.


## DATA CLEANING

**1. Delete unneccessary features by making copy of dataframe with neccessary features**

In [28]:
keep=['CNT', 'ST04Q01', 'PV1MATH','ST29Q02', 'ST29Q05', 'ST29Q07', 'ST29Q08', 'ST42Q01',
      'ST42Q03', 'ST42Q05', 'ST42Q08','ST42Q10', 'ST42Q02', 'ST42Q04', 'ST42Q06',
      'ST42Q07', 'ST42Q09', 'ST29Q01', 'ST29Q03', 'ST29Q04', 'ST29Q06','ST46Q01',
      'ST46Q02', 'ST46Q03', 'ST46Q04', 'ST46Q05', 'ST46Q06', 'ST46Q07', 'ST46Q08',
      'ST46Q09','ST49Q01', 'ST49Q02', 'ST49Q03', 'ST49Q04', 'ST49Q05', 'ST49Q06',
      'ST49Q07', 'ST49Q09','ST35Q04', 'ST35Q05', 'ST35Q06']

In [29]:
df_clean=df[keep]

In [30]:
df_clean.head()

CNT ST04Q01   PV1MATH         ST29Q02         ST29Q05         ST29Q07  \
0  Albania  Female  406.8469  Strongly agree           Agree           Agree   
1  Albania  Female  486.1427  Strongly agree           Agree        Disagree   
2  Albania  Female  533.2684  Strongly agree  Strongly agree  Strongly agree   
3  Albania  Female  412.2215             NaN             NaN             NaN   
4  Albania  Female  381.9209  Strongly agree  Strongly agree  Strongly agree   

          ST29Q08         ST42Q01   ST42Q03         ST42Q05   ...    \
0  Strongly agree           Agree     Agree           Agree   ...     
1        Disagree             NaN       NaN             NaN   ...     
2  Strongly agree             NaN       NaN             NaN   ...     
3             NaN             NaN  Disagree           Agree   ...     
4  Strongly agree  Strongly agree     Agree  Strongly agree   ...     

                   ST49Q02                  ST49Q03          ST49Q04  \
0                Sometimes                Sometimes        Sometimes   
1                    Often  Always or almost always        Sometimes   
2  Always or almost always                Sometimes  Never or rarely   
3                      NaN                      NaN              NaN   
4  Always or almost always                    Often            Often   

                   ST49Q05          ST49Q06          ST49Q07          ST49Q09  \
0                Sometimes  Never or rarely  Never or rarely  Never or rarely   
1  Always or almost always  Never or rarely  Never or rarely            Often   
2  Always or almost always  Never or rarely  Never or rarely  Never or rarely   
3                      NaN              NaN              NaN              NaN   
4                Sometimes              NaN        Sometimes        Sometimes   

          ST35Q04         ST35Q05 ST35Q06  
0           Agree           Agree   Agree  
1           Agree        Disagree   Agree  
2  Strongly agree  Strongly agree   Agree  
3             NaN             NaN     NaN  
4  Strongly agree  Strongly agree   Agree  

[5 rows x 41 columns]

**2. Combine the features in each section into one feature by assigning point values and taking average.**

(If student did not answer at least one feature under each section, then fill in Null Value for section's overall score)

Sections
+ Instrumental Motivation 
     + worthwhile for work, worthwhile for career changes, important for future study
        
        
+ Math Anxiety 
     + worry that it will be difficult, get very tense, get very nervous, feel helpless, worry about getting poor grades
        
        
+ Math Self-Concept 
     + not good at Math, get good grades, learn quickly, one of best subjects, understand difficult work
      
      
+ Math Interest 
     + enjoy reading, look forward to lessons, enjoy Math, interested in Math
        
        
+ Math Work Ethic 
     + homework completed on time, work hard on homework, prepared for exams, study hard for quizzes, study until I understand everything, pay attention in classes, listen in classes, avoid distractions when studying, keep work organized
        
        
+ Math Behaviour 
     + talk about Math with friends, help friends with Math, Extracurricular Activity, participate in competitions, study more than 2 extra hours a day, play chess, computer programming, participate in math club


+ Student's Parental View of Mathematics
    + parents believe studying mathematics is important, parents believe mathematics is important for career, parents like mathematics


**Sections Instrumental Motivation, Math Anxiety, Math Interest, Math Work Ethic, Student's Parental Attitude of Mathematics will be using the same scale where 'Strongly agree' would be 4 points and 'Strongly disagree' would be 1 point**



In [31]:
# Sections Instrumental Motivation, Math Anxiety, Math Interest, and Math Work Ethic
# will be using the same scale
points_1={'Strongly agree':4,'Agree': 3,'Disagree': 2, 'Strongly disagree':1}

In [32]:
# Function that will assign point values to each feature category
# and create/combine into one feature
# features variabe will be a list argument
# cmb_name variable will be the name of the new combined feature (string form)
# scale variable is what dictionary point system we will be using

def cmb_features(features,cmb_name, scale):
    for feature in features:
    
        df_clean.loc[:,feature]=df_clean.loc[:,feature].map(scale)

    df_clean[cmb_name]=df_clean.loc[:,features].sum(axis=1, skipna=False)/len(features)

In [33]:
# Instrumental Motivation
cmb_features(motivation,'motivation',points_1)

# Math Anxiety
cmb_features(anxiety, 'anxiety', points_1)

# Math Interest
cmb_features(interest, 'interest', points_1)

# Math Work Ethic
cmb_features(work_ethic, 'work_ethic', points_1)

cmb_features(parents, 'parents', points_1)

C:\Users\jsamala\Anaconda3\lib\site-packages\pandas\core\indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s
C:\Users\jsamala\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [34]:
df_clean[['motivation','anxiety','interest','work_ethic', 'parents']].head(10)

motivation  anxiety  interest  work_ethic   parents
0        3.50      2.6      3.00    3.000000  3.000000
1        2.75      NaN      2.25    3.222222  2.666667
2        4.00      NaN      3.25    3.888889  3.666667
3         NaN      NaN       NaN         NaN       NaN
4        4.00      3.2      2.50    3.777778  3.666667
5         NaN      2.2       NaN         NaN       NaN
6         NaN      2.8       NaN         NaN       NaN
7        3.50      2.4      3.00    3.666667  3.333333
8        3.00      2.4      3.00    3.000000  3.333333
9         NaN      1.4       NaN         NaN       NaN

**Section Math Behaviour would have a different points system due to being categorized differently. Features under this section contains categories Always or Almost Always, Often, Sometimes, Never or Rarely. In this scenario, we would have 'Always or Almost Always' assigned to 4 points and 'Often' to 3 points, and so on.**


In [35]:
# points system for Math Behavious
points_2={'Always or almost always': 4, 'Often':3, 'Sometimes':2, 'Never or rarely':1}

In [36]:
cmb_features(behavior,'behavior',points_2)

C:\Users\jsamala\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [37]:
df_clean['behavior'].head()

0    1.75
1    2.50
2    2.00
3     NaN
4     NaN
Name: behavior, dtype: float64

In [38]:
df_clean[behavior].head()

ST49Q01  ST49Q02  ST49Q03  ST49Q04  ST49Q05  ST49Q06  ST49Q07  ST49Q09
0      3.0      2.0      2.0      2.0      2.0      1.0      1.0      1.0
1      2.0      3.0      4.0      2.0      4.0      1.0      1.0      3.0
2      2.0      4.0      2.0      1.0      4.0      1.0      1.0      1.0
3      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
4      4.0      4.0      3.0      3.0      2.0      NaN      2.0      2.0


**Section Math Self-Concept would be an exception from the above scale system due to one feature asks students if they believe they are not good at Math meaning 'Strongly agree' would be a negative view of themselves. The other features under Math Self Concept have 'Strongly agree' as a positive view of themselves. This would be we would reverse the point system for just the feature 'Not Good at Maths' (ST42Q02)**



In [39]:
# reverse point system for 'Not Good at Math' feature
df_clean.loc[:,'ST42Q02']=df_clean.loc[:,'ST42Q02'].map({'Strongly agree':1, 
                                             'Agree': 2,'Disagree': 3, 'Strongly disagree':4})


In [40]:
# regular point system

for _ in ['ST42Q04','ST42Q06','ST42Q07','ST42Q09']:
    df_clean.loc[:,_]=df_clean.loc[:,_].map({'Strongly agree':4,'Agree': 3,'Disagree': 2, 'Strongly disagree':1})

In [41]:
df_clean[self].head()

ST42Q02  ST42Q04  ST42Q06  ST42Q07  ST42Q09
0      3.0      3.0      3.0      3.0      2.0
1      NaN      NaN      NaN      NaN      NaN
2      NaN      NaN      NaN      NaN      NaN
3      1.0      3.0      2.0      4.0      3.0
4      1.0      4.0      2.0      2.0      3.0

In [42]:
#create new feature that takes average of Self Features
df_clean['self']=df_clean.loc[:,self].sum(axis=1, skipna=False)/len(self)

C:\Users\jsamala\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [43]:
df_clean['self'].head()

0    2.8
1    NaN
2    NaN
3    2.6
4    2.4
Name: self, dtype: float64

In [44]:
df_clean[self].head()

ST42Q02  ST42Q04  ST42Q06  ST42Q07  ST42Q09
0      3.0      3.0      3.0      3.0      2.0
1      NaN      NaN      NaN      NaN      NaN
2      NaN      NaN      NaN      NaN      NaN
3      1.0      3.0      2.0      4.0      3.0
4      1.0      4.0      2.0      2.0      3.0

**3. Delete unneccessary features**

In [45]:
df_clean.drop(motivation+anxiety+self+interest+work_ethic+behavior+parents, axis=1,inplace=True)
df_clean.head()

C:\Users\jsamala\Anaconda3\lib\site-packages\pandas\core\frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


CNT ST04Q01   PV1MATH  motivation  anxiety  interest  work_ethic  \
0  Albania  Female  406.8469        3.50      2.6      3.00    3.000000   
1  Albania  Female  486.1427        2.75      NaN      2.25    3.222222   
2  Albania  Female  533.2684        4.00      NaN      3.25    3.888889   
3  Albania  Female  412.2215         NaN      NaN       NaN         NaN   
4  Albania  Female  381.9209        4.00      3.2      2.50    3.777778   

    parents  behavior  self  
0  3.000000      1.75   2.8  
1  2.666667      2.50   NaN  
2  3.666667      2.00   NaN  
3       NaN       NaN   2.6  
4  3.666667       NaN   2.4

**4. Rename feature labels**

In [46]:
col_names={'CNT':'country', 'ST04Q01':'gender', 'PV1MATH':'score_overall','PV1MACC':'score_1','PV1MACQ':'score_2','PV1MACS':'score_3','PV1MACU':'score_4','PV1MAPE':'score_5',
      'PV1MAPF':'score_6','PV1MAPI':'score_7'}

In [47]:
df_clean.rename(columns=col_names,inplace=True)

C:\Users\jsamala\Anaconda3\lib\site-packages\pandas\core\frame.py:3781: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


In [48]:
df_clean.head()

country  gender  score_overall  motivation  anxiety  interest  work_ethic  \
0  Albania  Female       406.8469        3.50      2.6      3.00    3.000000   
1  Albania  Female       486.1427        2.75      NaN      2.25    3.222222   
2  Albania  Female       533.2684        4.00      NaN      3.25    3.888889   
3  Albania  Female       412.2215         NaN      NaN       NaN         NaN   
4  Albania  Female       381.9209        4.00      3.2      2.50    3.777778   

    parents  behavior  self  
0  3.000000      1.75   2.8  
1  2.666667      2.50   NaN  
2  3.666667      2.00   NaN  
3       NaN       NaN   2.6  
4  3.666667       NaN   2.4

**5. Correctly categorize countries under CNT (country): Replace values of 'Florida (USA)','Massachusetts (USA)', and 'Connecticut (USA)' with 'United States of America'**

In [49]:
df_clean.country.value_counts()[['United States of America',
                                 'Florida (USA)','Massachusetts (USA)',
                                 'Connecticut (USA)' ]]

United States of America    4978
Florida (USA)               1896
Massachusetts (USA)         1723
Connecticut (USA)           1697
Name: country, dtype: int64

In [50]:
states=['Florida (USA)','Massachusetts (USA)','Connecticut (USA)']

df_clean['country'].replace(states, 'United States of America', inplace=True)

C:\Users\jsamala\Anaconda3\lib\site-packages\pandas\core\generic.py:5890: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


In [51]:
df_clean.country.value_counts()[['United States of America',
                                 'Florida (USA)','Massachusetts (USA)',
                                 'Connecticut (USA)' ]]

C:\Users\jsamala\Anaconda3\lib\site-packages\pandas\core\series.py:851: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  return self.loc[key]


United States of America    10294.0
Florida (USA)                   NaN
Massachusetts (USA)             NaN
Connecticut (USA)               NaN
Name: country, dtype: float64

In [52]:
df_clean.head()

country  gender  score_overall  motivation  anxiety  interest  work_ethic  \
0  Albania  Female       406.8469        3.50      2.6      3.00    3.000000   
1  Albania  Female       486.1427        2.75      NaN      2.25    3.222222   
2  Albania  Female       533.2684        4.00      NaN      3.25    3.888889   
3  Albania  Female       412.2215         NaN      NaN       NaN         NaN   
4  Albania  Female       381.9209        4.00      3.2      2.50    3.777778   

    parents  behavior  self  
0  3.000000      1.75   2.8  
1  2.666667      2.50   NaN  
2  3.666667      2.00   NaN  
3       NaN       NaN   2.6  
4  3.666667       NaN   2.4

## Complete! Our dataset is looking good! :)

In [53]:
df_clean.to_csv('pisa2012_clean.csv', index_label=False)